File: ios_tracker/BallTracker/Detection/BallDetector.swift
import CoreImage
import CoreML
import CoreVideo
import Foundation
import ImageIO
import UniformTypeIdentifiers

/// A decoded ball detection in source-frame pixel space.
struct BallDetection {
    let box: CGRect
    let conf: Float

    var center: CGPoint { CGPoint(x: box.midX, y: box.midY) }
}

enum BallDetectorError: Error {
    case modelMissing
    case letterboxFailed
    case badOutput
}

/// Runs the NMS-free ball_best Core ML model (raw [1,5,10710] output) fully on
/// the Neural Engine, then decodes + NMS on the CPU. Keeping NMS out of the
/// graph is deliberate: baked NonMaxSuppression/TopK ops are not
/// ANE-supported and force graph partitioning (see spikes/FINDINGS.md), and it
/// keeps the ultra-low conf thresholds the tennis pipeline relies on tunable
/// per call-site.
final class BallDetector {
    static let inputWidth = 960
    static let inputHeight = 544

    /// ACTIVE_BALL_CONF from the pipeline — the operating threshold for the
    /// online tracker, which must commit to an association every frame and so
    /// needs precision over recall.
    static let defaultConf: Float = 0.10

    /// Threshold for the offline Viterbi solver, which wants the opposite
    /// trade: it discriminates by trajectory dynamics, so it would rather sift
    /// a pile of weak candidates than never see the ball. Measured on real
    /// footage, at 0.10 the ball is missed in 48% of frames; at 0.03, 4%.
    /// Only safe because exclusion zones already fence off stationary clutter.
    static let solverConf: Float = 0.10
    /// Ultralytics' default NMS IoU, matching the parity fixtures.
    static let defaultIoU: Float = 0.7

    /// Upper bound on a ball box, in model-input px (960-wide space).
    ///
    /// Deliberately loose. It is tempting to derive this from geometry — a
    /// 6.7 cm ball at 18–93 ft through this lens is only ~1–6 px — but the
    /// model boxes the *motion streak*, not the ball: at 30 fps a 60 mph ball
    /// smears ~26 px across a frame. Measured on real footage, real-ball boxes
    /// run 9–38 px wide and barely vary with distance (1.4× top-to-bottom,
    /// where geometry demands ~5×), so box size tracks ball *speed*, not range.
    /// A physical bound would reject every real ball. This only fences off
    /// grossly non-ball blobs (players, bags, chairs).
    static let defaultMaxBoxPx: Float = 45

    /// The Simulator has no ANE and its MPSGraph backend doesn't reliably
    /// support this model's ops, so `.all` there throws an internal Espresso
    /// exception (MpsGraph backend validation on incompatible OS) that can
    /// leave the prediction output torn and crash on decode. CPU-only is the
    /// only compute path the Simulator can run this model on.
    static let defaultComputeUnits: MLComputeUnits = {
        #if targetEnvironment(simulator)
        return .cpuOnly
        #else
        return .all
        #endif
    }()

    private let model: MLModel
    private let letterbox: Letterbox
    private let inputName = "image"
    private let outputName = "detections"

    /// When BALL_DEBUG_DUMP is set in the environment, print the top raw
    /// confidence per frame (pre-threshold) and save the first few letterboxed
    /// 960×544 model-input frames as PNGs, so the exact on-device input can be
    /// diffed against the Python/Ultralytics letterbox.
    private let debugDump = ProcessInfo.processInfo.environment["BALL_DEBUG_DUMP"] != nil
    private var dumpCount = 0
    private let dumpMax = 5
    private let debugCIContext = CIContext(options: [.cacheIntermediates: false])

    convenience init(computeUnits: MLComputeUnits = BallDetector.defaultComputeUnits) throws {
        guard let url = Bundle.main.url(forResource: "ball_best", withExtension: "mlmodelc") else {
            throw BallDetectorError.modelMissing
        }
        try self.init(modelURL: url, computeUnits: computeUnits)
    }

    init(modelURL: URL, computeUnits: MLComputeUnits = BallDetector.defaultComputeUnits) throws {
        guard let letterbox = Letterbox(width: Self.inputWidth, height: Self.inputHeight) else {
            throw BallDetectorError.letterboxFailed
        }
        let cfg = MLModelConfiguration()
        cfg.computeUnits = computeUnits
        self.model = try MLModel(contentsOf: modelURL, configuration: cfg)
        self.letterbox = letterbox
    }

    /// Detect balls in a frame of any size; boxes come back in that frame's
    /// pixel space.
    func detect(in pixelBuffer: CVPixelBuffer,
                conf: Float = BallDetector.defaultConf,
                iouThreshold: Float = BallDetector.defaultIoU,
                maxBoxPx: Float = BallDetector.defaultMaxBoxPx) throws -> [BallDetection] {
        guard let (input, transform) = letterbox.apply(to: pixelBuffer) else {
            throw BallDetectorError.letterboxFailed
        }
        if debugDump { dumpLetterboxedInput(input) }
        let provider = try MLDictionaryFeatureProvider(
            dictionary: [inputName: MLFeatureValue(pixelBuffer: input)])
        let output = try model.prediction(from: provider)
        guard let arr = output.featureValue(for: outputName)?.multiArrayValue else {
            throw BallDetectorError.badOutput
        }
        return decode(arr, conf: conf, iouThreshold: iouThreshold,
                      maxBoxPx: maxBoxPx, transform: transform)
    }

    /// Saves the letterboxed model input as a PNG and logs where, so it can be
    /// pulled off the device/simulator and re-run through Python.
    private func dumpLetterboxedInput(_ input: CVPixelBuffer) {
        guard dumpCount < dumpMax else { return }
        let idx = dumpCount
        dumpCount += 1
        let ci = CIImage(cvPixelBuffer: input)
        guard let cg = debugCIContext.createCGImage(ci, from: ci.extent) else { return }
        let dir = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask)[0]
        let url = dir.appendingPathComponent("letterbox_\(idx).png")
        guard let dest = CGImageDestinationCreateWithURL(
            url as CFURL, UTType.png.identifier as CFString, 1, nil) else { return }
        CGImageDestinationAddImage(dest, cg, nil)
        if CGImageDestinationFinalize(dest) {
            print("[BallDetector] dumped letterboxed input -> \(url.path)")
        }
    }

    /// Decode the raw [1,5,N] head (cx,cy,w,h,conf per anchor, channel-major),
    /// threshold, NMS, and un-letterbox to source pixels.
    private func decode(_ arr: MLMultiArray, conf: Float, iouThreshold: Float,
                        maxBoxPx: Float, transform: LetterboxTransform) -> [BallDetection] {
        guard arr.shape.count == 3, arr.shape[1].intValue == 5 else { return [] }
        let n = arr.shape[2].intValue
        // The backing buffer can be padded (e.g. channel stride 10720 for
        // n=10710), so index via the declared strides, never densely.
        let declaredChStride = arr.strides[1].intValue
        let declaredAnchorStride = arr.strides[2].intValue

        var candidates: [(box: CGRect, conf: Float)] = []
        var maxConf: Float = 0
        arr.withUnsafeBufferPointer(ofType: Float.self) { p in
            // On device (and macOS, .all/.cpuOnly) the backing buffer is
            // allocated at the *padded* size, so the declared strides read in
            // bounds. The iOS Simulator's CPU/MPS backend reports the same
            // padded strides but hands back a buffer only the *dense* size —
            // stride math then runs past the allocation and crashes with
            // EXC_BAD_ACCESS on decode. Only trust the declared strides while
            // the highest index they produce stays in bounds; otherwise fall
            // back to dense channel-major strides ([5n, n, 1]).
            let maxDeclaredIdx = (n - 1) * declaredAnchorStride + 4 * declaredChStride
            let stridesInBounds = maxDeclaredIdx < p.count
            let chStride = stridesInBounds ? declaredChStride : n
            let anchorStride = stridesInBounds ? declaredAnchorStride : 1
            for i in 0..<n {
                let base = i * anchorStride
                let c = p[base + 4 * chStride]
                if c > maxConf { maxConf = c }
                if c < conf { continue }
                let cx = CGFloat(p[base])
                let cy = CGFloat(p[base + chStride])
                let w = CGFloat(p[base + 2 * chStride])
                let h = CGFloat(p[base + 3 * chStride])
                if Float(w) > maxBoxPx || Float(h) > maxBoxPx { continue }
                candidates.append((CGRect(x: cx - w / 2, y: cy - h / 2, width: w, height: h), c))
            }
        }

        if debugDump {
            print(String(format: "[BallDetector] maxRawConf=%.4f  above %.2f=%d",
                         maxConf, conf, candidates.count))
        }

        let kept = nms(candidates, iouThreshold: iouThreshold)
        return kept.map { cand in
            let o = transform.unmap(cand.box.origin)
            let box = CGRect(x: o.x, y: o.y,
                             width: cand.box.width / transform.scale,
                             height: cand.box.height / transform.scale)
            return BallDetection(box: box, conf: cand.conf)
        }
    }

    /// Greedy single-class non-maximum suppression, highest confidence first.
    private func nms(_ candidates: [(box: CGRect, conf: Float)],
                     iouThreshold: Float) -> [(box: CGRect, conf: Float)] {
        guard candidates.count > 1 else { return candidates }
        let sorted = candidates.sorted { $0.conf > $1.conf }
        var kept: [(box: CGRect, conf: Float)] = []
        for cand in sorted {
            var suppressed = false
            for k in kept where iou(cand.box, k.box) > iouThreshold {
                suppressed = true
                break
            }
            if !suppressed { kept.append(cand) }
        }
        return kept
    }

    private func iou(_ a: CGRect, _ b: CGRect) -> Float {
        let inter = a.intersection(b)
        if inter.isNull || inter.isEmpty { return 0 }
        let interArea = inter.width * inter.height
        let union = a.width * a.height + b.width * b.height - interArea
        return union > 0 ? Float(interArea / union) : 0
    }
}

File: android_tracker/app/src/main/java/com/build2launch/balltracker/detection/BallDetector.kt
package com.build2launch.balltracker.detection

import android.content.Context
import android.graphics.Bitmap
import android.graphics.RectF
import android.os.Build
import android.util.Log
import org.tensorflow.lite.Interpreter
import org.tensorflow.lite.gpu.CompatibilityList
import org.tensorflow.lite.gpu.GpuDelegate
import org.tensorflow.lite.nnapi.NnApiDelegate
import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.nio.channels.FileChannel

/** A decoded ball detection in source-frame pixel space. */
data class BallDetection(val box: RectF, val conf: Float) {
    val centerX: Float get() = box.centerX()
    val centerY: Float get() = box.centerY()
}

/**
 * Runs the NMS-free ball_best model on Android's built-in ML acceleration —
 * TensorFlow Lite (LiteRT) with the NNAPI delegate (NPU/DSP), falling back to
 * the GPU delegate, then multithreaded CPU (XNNPACK). Decode + single-class
 * NMS run on the CPU afterwards, exactly like the iOS BallDetector: keeping
 * NMS out of the graph keeps the whole conv net on the accelerator and the
 * ultra-low conf thresholds the tennis pipeline relies on tunable per call.
 *
 * Port of ios_tracker BallTracker/Detection/BallDetector.swift.
 */
class BallDetector private constructor(
    private val interpreter: Interpreter,
    private val nnApiDelegate: NnApiDelegate?,
    private val gpuDelegate: GpuDelegate?,
    /** The accelerator that actually backed this interpreter, for the HUD/logs. */
    val backend: String,
) : AutoCloseable {

    private val letterbox = Letterbox(inputWidth, inputHeight)

    // Output tensor geometry, read from the model at load time so the decode
    // adapts to whichever layout the exporter produced.
    private val outShape: IntArray = interpreter.getOutputTensor(0).shape()
    private val channelMajor: Boolean   // true: [1,5,N]; false: [1,N,5]
    private val numAnchors: Int
    private val outputBuffer: ByteBuffer

    /**
     * Ultralytics' TFLite export emits box coords normalized to [0,1]; the .pt
     * / CoreML paths emit pixels. Auto-detected on the first frame that carries
     * a real box and then cached, so parity holds whichever export is bundled.
     */
    private var coordsArePixels: Boolean? = null

    init {
        // Expect [1, 5, N] or [1, N, 5]; channel count is whichever axis is 5.
        require(outShape.size == 3 && outShape[0] == 1) {
            "unexpected output shape ${outShape.joinToString()}"
        }
        if (outShape[1] == 5) {
            channelMajor = true
            numAnchors = outShape[2]
        } else {
            channelMajor = false
            numAnchors = outShape[1]
        }
        outputBuffer = ByteBuffer
            .allocateDirect(outShape[1] * outShape[2] * 4)
            .order(ByteOrder.nativeOrder())
    }

    /**
     * Detect balls in a bitmap of any size; boxes come back in that bitmap's
     * pixel space. Thread-confined: call from a single inference thread.
     */
    fun detect(
        bitmap: Bitmap,
        conf: Float = defaultConf,
        iouThreshold: Float = defaultIoU,
        maxBoxPx: Float = defaultMaxBoxPx,
    ): List<BallDetection> {
        val transform = letterbox.apply(bitmap)
        outputBuffer.rewind()
        interpreter.run(letterbox.inputBuffer, outputBuffer)
        outputBuffer.rewind()
        return decode(outputBuffer.asFloatBuffer(), conf, iouThreshold, maxBoxPx, transform)
    }

    /** Value of channel `ch` for anchor `i` in the flat output buffer. */
    private fun get(fb: java.nio.FloatBuffer, ch: Int, i: Int): Float =
        if (channelMajor) fb.get(ch * numAnchors + i) else fb.get(i * 5 + ch)

    private fun decode(
        fb: java.nio.FloatBuffer,
        conf: Float,
        iouThreshold: Float,
        maxBoxPx: Float,
        transform: LetterboxTransform,
    ): List<BallDetection> {
        // Resolve pixel-vs-normalized coords once, from the strongest anchor.
        if (coordsArePixels == null) {
            var maxCoord = 0f
            var i = 0
            while (i < numAnchors) {
                val cx = get(fb, 0, i); val cy = get(fb, 1, i)
                if (cx > maxCoord) maxCoord = cx
                if (cy > maxCoord) maxCoord = cy
                i++
            }
            // Normalized coords live in [0,1]; a few px of slop → threshold 4.
            coordsArePixels = maxCoord > 4f
        }
        val sx = if (coordsArePixels == true) 1f else inputWidth.toFloat()
        val sy = if (coordsArePixels == true) 1f else inputHeight.toFloat()

        data class Cand(val box: RectF, val conf: Float)
        val candidates = ArrayList<Cand>()
        var i = 0
        while (i < numAnchors) {
            val c = get(fb, 4, i)
            if (c >= conf) {
                val cx = get(fb, 0, i) * sx
                val cy = get(fb, 1, i) * sy
                val w = get(fb, 2, i) * sx
                val h = get(fb, 3, i) * sy
                if (w <= maxBoxPx && h <= maxBoxPx) {
                    candidates.add(Cand(RectF(cx - w / 2, cy - h / 2, cx + w / 2, cy + h / 2), c))
                }
            }
            i++
        }

        val kept = nms(candidates.map { it.box to it.conf }, iouThreshold)
        return kept.map { (box, c) ->
            val ox = transform.unmapX(box.left)
            val oy = transform.unmapY(box.top)
            val bw = box.width() / transform.scale
            val bh = box.height() / transform.scale
            BallDetection(RectF(ox, oy, ox + bw, oy + bh), c)
        }
    }

    /** Greedy single-class non-maximum suppression, highest confidence first. */
    private fun nms(cands: List<Pair<RectF, Float>>, iouThreshold: Float): List<Pair<RectF, Float>> {
        if (cands.size <= 1) return cands
        val sorted = cands.sortedByDescending { it.second }
        val kept = ArrayList<Pair<RectF, Float>>()
        for (cand in sorted) {
            var suppressed = false
            for (k in kept) {
                if (iou(cand.first, k.first) > iouThreshold) { suppressed = true; break }
            }
            if (!suppressed) kept.add(cand)
        }
        return kept
    }

    private fun iou(a: RectF, b: RectF): Float {
        val ix = maxOf(a.left, b.left)
        val iy = maxOf(a.top, b.top)
        val ax = minOf(a.right, b.right)
        val ay = minOf(a.bottom, b.bottom)
        val iw = ax - ix
        val ih = ay - iy
        if (iw <= 0 || ih <= 0) return 0f
        val inter = iw * ih
        val union = a.width() * a.height() + b.width() * b.height() - inter
        return if (union > 0) inter / union else 0f
    }

    override fun close() {
        interpreter.close()
        nnApiDelegate?.close()
        gpuDelegate?.close()
    }

    companion object {
        private const val TAG = "BallDetector"
        const val inputWidth = 960
        const val inputHeight = 544

        /** ACTIVE_BALL_CONF — the online tracker's operating threshold. */
        const val defaultConf = 0.10f
        /** Offline Viterbi solver threshold (recall over precision). */
        const val solverConf = 0.10f
        /** Ultralytics' default NMS IoU, matching the parity fixtures. */
        const val defaultIoU = 0.7f
        /** Upper bound on a ball box, in model-input px (see iOS notes). */
        const val defaultMaxBoxPx = 45f

        const val ASSET_NAME = "ball_best.tflite"

        /** Load the bundled model, choosing the best available accelerator. */
        fun fromAsset(context: Context, assetName: String = ASSET_NAME): BallDetector {
            val model = loadModelBuffer(context, assetName)
            return build(model)
        }

        private fun build(model: ByteBuffer): BallDetector {
            // 1. NNAPI — Android's built-in NPU/DSP path (API 27+).
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O_MR1) {
                try {
                    val delegate = NnApiDelegate()
                    val opts = Interpreter.Options().addDelegate(delegate)
                    val interp = Interpreter(model, opts)
                    Log.i(TAG, "using NNAPI delegate")
                    return BallDetector(interp, delegate, null, "NNAPI")
                } catch (t: Throwable) {
                    Log.w(TAG, "NNAPI delegate unavailable: ${t.message}")
                }
            }
            // 2. GPU delegate.
            try {
                if (CompatibilityList().isDelegateSupportedOnThisDevice) {
                    val delegate = GpuDelegate()
                    val opts = Interpreter.Options().addDelegate(delegate)
                    val interp = Interpreter(model, opts)
                    Log.i(TAG, "using GPU delegate")
                    return BallDetector(interp, null, delegate, "GPU")
                }
            } catch (t: Throwable) {
                Log.w(TAG, "GPU delegate unavailable: ${t.message}")
            }
            // 3. CPU (XNNPACK, multithreaded).
            val opts = Interpreter.Options().apply {
                numThreads = Runtime.getRuntime().availableProcessors().coerceAtMost(4)
                useXNNPACK = true
            }
            Log.i(TAG, "using CPU (XNNPACK)")
            return BallDetector(Interpreter(model, opts), null, null, "CPU")
        }

        private fun loadModelBuffer(context: Context, assetName: String): ByteBuffer {
            context.assets.openFd(assetName).use { fd ->
                fd.createInputStream().channel.use { channel ->
                    return channel.map(
                        FileChannel.MapMode.READ_ONLY,
                        fd.startOffset,
                        fd.declaredLength,
                    )
                }
            }
        }
    }
}
